distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso

```diff
@@ -1169,3 +1189,113 @@ type Incomparable_Values_Error
 handle_incomparable_value ~function =
     handle t = Panic.catch t handler=(Error.throw Incomparable_Values_Error)
     handle No_Such_Method_Error <| handle Type_Error <| handle Unsupported_Argument_Types <| function
+
+## PRIVATE
+   Resolves a vector of ranges or indices into a vector of ranges that fit
+   within a sequence.
+resolve_ranges : Vector (Integer | Range) -> Integer -> Vector Range
+resolve_ranges ranges length =
+    ## Ensures that a descriptor fits within the range of the current
+       vector, trimming it or reporting an error if it is invalid.
+    trim descriptor = case descriptor of
+        Integer ->
+            actual_index = if descriptor < 0 then length + descriptor else descriptor
+            if (actual_index < 0) || (actual_index >= length) then Panic.throw (Index_Out_Of_Bounds_Error descriptor length) else
+                actual_index
+        Range start end step ->
+            if step <= 0 then Panic.throw (Illegal_Argument_Error "Range step must be positive.") else
+                if (start < 0) || (end < 0) then Panic.throw (Illegal_Argument_Error "Range start and end must not be negative.") else
+                    if start >= length then Panic.throw (Index_Out_Of_Bounds_Error start length) else
+                        actual_end = Math.min end length
+                        if actual_end < start then Range start start step else
+                            Range start actual_end step
+    ranges.map trim
+
+## PRIVATE
+   Creates a new vector where for each range, a corresponding section of the
+   source vector is added to the result.
+
+   Assumes that the ranges have been already bounds-checked (for example by
+   passing them through `resolve_ranges`).
+slice_ranges vector ranges =
+    if ranges.length == 0 then empty else case ranges.length == 1 of
+        True -> case ranges.first of
+            Integer -> Vector (Array.new_1 (vector.unsafe_at ranges.first))
```

Comment:
I added this special case, because it can be triggered _relatively_ often by `vec.take (By_Index 123)`. It seemed that creating a slice just to hold a single element will be an overkill - because the cost of allocating the atom representing the slice descriptor should be roughly the same as the cost of allocating a single-element array and the memory footprint of both of these is also comparable, so it seems to me that making a slice for just a single element would be wasteful.

Reply:
We could do a faster short circuit earlier that if By_Index is an Integer or a single Integer Vector we return just `self.at index` and let the self.at do the checking and resolving,

Reply:
We could, but then again I still need to do the check at `slice_ranges` for a single range - so I still need this case-of but now it will be in two places.